File: skills/ablate/tests/usage-counts.test.ts
/// <reference types="node" />
// In-process tests for skills/ablate/scripts/usage_counts.ts, the TS port of
// usage_counts.py's FIRE_EVENTS, ELEMENT_SUFFIXES, RARE_BY_DESIGN, MEASUREMENT_WINDOW_DAYS,
// element_path, _parse_date, _iter_fires, count_usage, classify and main.
// skills/ablate/tests/usage-counts-fixture.test.ts already froze
// skills/ablate/tests/fixtures/usage-counts-cases.json against the real python3
// usage_counts.py; T-354 below replays that same frozen fixture against the .ts port's
// count_usage directly, rather than hand-writing a second set of expected rows that could
// drift from it (docs/wiki/fixture-freeze-before-port.md), the same DRY choice
// skills/ablate/tests/enforcer-map.test.ts makes for enforcer_map.ts.
import assert from "node:assert/strict";
import { readFileSync, rmSync } from "node:fs";
import { dirname, join } from "node:path";
import test from "node:test";
import { fileURLToPath } from "node:url";
import { fixture } from "../../../workflows/_lib/tests/_cli-fixture.ts";
import { writeTree } from "../../_lib/tests/_python-cli-fixture.ts";
import { DELETE_CANDIDATE, NEEDS_HUMAN_JUDGMENT } from "../scripts/verdict.ts";
import { UNMEASURED } from "../scripts/arms.ts";
import {
  classify,
  count_usage,
  MEASUREMENT_WINDOW_DAYS,
  RARE_BY_DESIGN,
} from "../scripts/usage_counts.ts";
import type { UsageResult } from "../scripts/usage_counts.ts";

const HERE = dirname(fileURLToPath(import.meta.url));

interface FixtureCase {
  name: string;
  files: Record<string, string>;
  argv: string[];
  exit: number;
  stdout: string;
}

const CASES = JSON.parse(
  readFileSync(join(HERE, "fixtures", "usage-counts-cases.json"), "utf8"),
) as FixtureCase[];

test("T-354 count_usage tallies the fires inside the window and drops the ones outside it, driven by the frozen transcript", () => {
  // "the window" here is the set FIRE_EVENTS/ELEMENT_SUFFIXES recognize as one hook fire
  // naming a harness element: counts_fires_across_transcripts_and_drops_a_label_only_command
  // is the frozen fixture case that puts an element-shaped fire ("~/.claude/hooks/.../*.py")
  // and a label-only fire ("formatter", which names no element and so falls outside that set)
  // in the same transcript pair. Replaying it straight through the ported count_usage proves
  // the label-only one is dropped rather than tallied, and the element-shaped ones are.
  const entry = fixture(CASES, "counts_fires_across_transcripts_and_drops_a_label_only_command");
  const root = writeTree("usage-counts-count", entry.files);
  try {
    const expected = JSON.parse(entry.stdout) as UsageResult;
    const result = count_usage(root);
    assert.deepEqual(result, expected);
  } finally {
    rmSync(root, { recursive: true, force: true });
  }
});

test("T-355 a path in RARE_BY_DESIGN classifies as measured rather than as a delete candidate", () => {
  const [rarePath] = RARE_BY_DESIGN;
  const now = new Date("2026-08-27T00:00:00.000Z");

  // Zero fires and no last_used would read as unused for any other path -- RARE_BY_DESIGN's
  // first-row-wins check must still keep this one out of DELETE_CANDIDATE.
  const verdict = classify(rarePath, 0, null, now, MEASUREMENT_WINDOW_DAYS);

  assert.notEqual(verdict, DELETE_CANDIDATE);
  assert.equal(verdict, NEEDS_HUMAN_JUDGMENT);
});

test("T-356 classify takes the window as an argument, so a caller can drive both sides of the boundary without patching a module binding", () => {
  const path = "hooks/pre-bash/wiki_scene.py";
  const now = new Date("2026-08-27T00:00:00.000Z");
  // 238 days before `now` (usage_counts_test.py's MeasurementWindow case uses the same shape:
  // a last_used date that a narrow window reports stale and a wide window reports current).
  const lastUsed = "2026-01-01";

  assert.equal(classify(path, 5, lastUsed, now, 30), UNMEASURED);
  assert.equal(classify(path, 5, lastUsed, now, 300), NEEDS_HUMAN_JUDGMENT);
});

File: skills/ablate/tests/usage_counts_test.py
"""Tests for skills/ablate/scripts/usage_counts.py.

Run: python3 skills/ablate/tests/usage_counts_test.py

Fixture records follow the shape a real ~/.claude/projects/**/*.jsonl transcript carries for
a hook fire (confirmed by reading a live transcript in this session): a top-level "timestamp"
paired with an "attachment" object whose "hookEvent" is PreToolUse or PostToolUse and whose
"command" names the hook script that fired. A real "command" carries the home-relative form
the harness invoked ("~/.claude/hooks/pre-bash/wiki_scene.ts"), measured over this session's
transcripts, so the fixtures below write that form rather than the repo-root-relative path a
harness element is named by.
"""

import json
import sys
import tempfile
import unittest
from datetime import date
from pathlib import Path, PurePosixPath
from unittest.mock import patch

HERE = Path(__file__).resolve().parent
sys.path.insert(0, str(HERE.parent / "scripts"))

import usage_counts  # noqa: E402
from arms import UNMEASURED  # noqa: E402
from verdict import DELETE_CANDIDATE  # noqa: E402


def _write_transcript(root: Path, rel: str, records: list[dict[str, object]]) -> Path:
    path = root / rel
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("w", encoding="utf-8") as fh:
        for record in records:
            fh.write(json.dumps(record) + "\n")
    return path


def _fire(*, event: str, command: str, timestamp: str) -> dict[str, object]:
    """One PreToolUse/PostToolUse attachment record."""
    return {
        "type": "attachment",
        "attachment": {
            "type": "hook_success",
            "hookName": f"{event}:Bash",
            "hookEvent": event,
            "command": command,
            "stdout": "",
            "exitCode": 0,
        },
        "timestamp": timestamp,
    }


class FireCounting(unittest.TestCase):
    def test_a_hook_fire_is_counted_from_a_pretooluse_record_in_a_session_transcript(
        self,
    ) -> None:
        """T-001 A hook fire is counted from a PreToolUse record in a session transcript"""
        with tempfile.TemporaryDirectory() as tmp:
            root = Path(tmp)
            _write_transcript(
                root,
                "project-a/session-1.jsonl",
                [
                    _fire(
                        event="PreToolUse",
                        command="~/.claude/hooks/pre-bash/wiki_scene.ts",
                        timestamp="2026-08-01T00:00:00.000Z",
                    )
                ],
            )

            result = usage_counts.count_usage(root)

            self.assertEqual(result["elements"]["hooks/pre-bash/wiki_scene.ts"]["fires"], 1)


class LabelFires(unittest.TestCase):
    def test_a_fire_whose_command_is_a_label_rather_than_a_path_is_left_out_of_the_tally(
        self,
    ) -> None:
        """T-007 A fire whose command is a label rather than a path is left out of the tally"""
        # "formatter" is a command value measured in real transcripts. It names no harness
        # element, so counting it would put a key in the tally that nothing can join to.
        with tempfile.TemporaryDirectory() as tmp:
            root = Path(tmp)
            _write_transcript(
                root,
                "project-a/session-1.jsonl",
                [
                    _fire(
                        event="PostToolUse",
                        command="formatter",
                        timestamp="2026-08-01T00:00:00.000Z",
                    ),
                    _fire(
                        event="PreToolUse",
                        command="~/.claude/hooks/pre-bash/wiki_scene.ts",
                        timestamp="2026-08-01T00:00:00.000Z",
                    ),
                ],
            )

            result = usage_counts.count_usage(root)

            self.assertEqual(list(result["elements"]), ["hooks/pre-bash/wiki_scene.ts"])


class TypeScriptElements(unittest.TestCase):
    def test_a_fire_naming_a_typescript_hook_is_counted_the_way_a_python_one_is(self) -> None:
        """A fire naming a TypeScript hook is counted the way a Python one is"""
        # RARE_BY_DESIGN names a .ts hook, and a suffix set without .ts drops that path before
        # the tally is built, leaving the entry pointing at a key nothing can produce.
        with tempfile.TemporaryDirectory() as tmp:
            root = Path(tmp)
            _write_transcript(
                root,
                "project-a/session-1.jsonl",
                [
                    _fire(
                        event="PreToolUse",
                        command="~/.claude/hooks/security/rm_to_trash.ts",
                        timestamp="2026-08-01T00:00:00.000Z",
                    )
                ],
            )

            result = usage_counts.count_usage(root)

            self.assertEqual(result["elements"]["hooks/security/rm_to_trash.ts"]["fires"], 1)

        for path in usage_counts.RARE_BY_DESIGN:
            self.assertIn(
                PurePosixPath(path).suffix,
                usage_counts.ELEMENT_SUFFIXES,
                f"{path} names a suffix the tally drops, so the entry can never match",
            )


class RareByDesign(unittest.TestCase):
    def test_an_element_flagged_rare_by_design_is_not_reported_as_a_delete_candidate_at_zero_fires(
        self,
    ) -> None:
        """T-002 An element flagged rare-by-design is not reported as a delete candidate at
        zero fires"""
        # Patched rather than read from whatever paths the module ships with, so shipping a
        # different set cannot silently turn this case into a no-op. A name that is not the
        # real RARE_BY_DESIGN entry, so this case cannot pass by accidentally matching it.
        rare_path = "hooks/security/example_guard.py"
        with patch.object(usage_counts, "RARE_BY_DESIGN", frozenset({rare_path})):
            verdict = usage_counts.classify(
                rare_path, fires=0, last_used=None, now=date(2026, 8, 27)
            )

        self.assertNotEqual(verdict, DELETE_CANDIDATE)


class MeasurementWindow(unittest.TestCase):
    def test_an_element_last_used_outside_the_measurement_window_is_reported_as_unmeasured(
        self,
    ) -> None:
        """T-003 An element last used outside the measurement window is reported as
        unmeasured"""
        # Patched rather than hand-picking a date that already exceeds the shipped default,
        # so moving the constant is what decides the verdict.
        with patch.object(usage_counts, "MEASUREMENT_WINDOW_DAYS", 30):
            stale_last_used = date(2026, 1, 1).isoformat()  # outside a 30-day window
            verdict = usage_counts.classify(
                "hooks/pre-bash/wiki_scene.ts",
                fires=5,
                last_used=stale_last_used,
                now=date(2026, 8, 27),
            )

        self.assertEqual(verdict, UNMEASURED)


class TranscriptSummary(unittest.TestCase):
    def test_the_output_carries_the_parsed_transcript_count_and_its_date_range(
        self,
    ) -> None:
        """T-004 The output carries the parsed transcript count and its date range"""
        with tempfile.TemporaryDirectory() as tmp:
            root = Path(tmp)
            _write_transcript(
                root,
                "project-a/session-1.jsonl",
                [
                    _fire(
                        event="PreToolUse",
                        command="~/.claude/hooks/pre-bash/wiki_scene.ts",
                        timestamp="2026-08-01T00:00:00.000Z",
                    )
                ],
            )
            _write_transcript(
                root,
                "project-b/session-2.jsonl",
                [
                    _fire(
                        event="PostToolUse",
                        command="~/.claude/hooks/post-bash/scribe_prompt.ts",
                        timestamp="2026-08-10T00:00:00.000Z",
                    )
                ],
            )

            result = usage_counts.count_usage(root)

            self.assertEqual(result["transcript_count"], 2)
            self.assertEqual(result["date_range"], {"start": "2026-08-01", "end": "2026-08-10"})


if __name__ == "__main__":
    _ = unittest.main(verbosity=2)
